From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ben Hutchings <ben@decadent.org.uk>
Date: Tue, 8 Jul 2025 13:32:21 +0200
Subject: Export symbols needed by binder

We want to enable use of the Android binder driver to support
Waydroid, but it should not be built-in as that would waste resources
and increase security attack surface on systems that don't need it.

Export the currently un-exported symbols it depends on.

Bug-Debian: https://bugs.debian.org/901492

Debian-patch: debian/patches/debian/export-symbols-needed-by-binder.patch (Debian linux 7.2.9-1, salsa.debian.org/kernel-team/linux tag debian/7.2.9-1)
---
 fs/file.c               | 1 +
 ipc/msgutil.c           | 1 +
 ipc/namespace.c         | 1 +
 kernel/sched/syscalls.c | 1 +
 kernel/sched/wait.c     | 1 +
 kernel/task_work.c      | 1 +
 mm/list_lru.c           | 1 +
 mm/memory.c             | 1 +
 mm/mmap_lock.c          | 1 +
 security/security.c     | 4 ++++
 10 files changed, 13 insertions(+)

diff --git a/fs/file.c b/fs/file.c
index 628ca07dc4b179db98f0ac6e9f6d60dd7f3c5d3b..f4287ef63a12d25d25c1c5a43de871347fe4ffdc 100644
--- a/fs/file.c
+++ b/fs/file.c
@@ -886,6 +886,7 @@ struct file *file_close_fd(unsigned int fd)
 
 	return file;
 }
+EXPORT_SYMBOL_GPL(file_close_fd);
 
 void do_close_on_exec(struct files_struct *files)
 {
diff --git a/ipc/msgutil.c b/ipc/msgutil.c
index e28f0cecb2ec942a4f6ee93df8384716bd026011..89a8fae5d90ce74d93cacedc08fe7db797eccd4d 100644
--- a/ipc/msgutil.c
+++ b/ipc/msgutil.c
@@ -35,6 +35,7 @@ struct msg_msgseg {
 	struct msg_msgseg *next;
 	/* the next part of the message follows immediately */
 };
+EXPORT_SYMBOL_GPL(init_ipc_ns);
 
 #define DATALEN_MSG	((size_t)PAGE_SIZE-sizeof(struct msg_msg))
 #define DATALEN_SEG	((size_t)PAGE_SIZE-sizeof(struct msg_msgseg))
diff --git a/ipc/namespace.c b/ipc/namespace.c
index 1e71353bdb4a330e84b3b35b08a2ba58351faaa5..93dd24802ce140a800f60f30d13d32a5d17dfd80 100644
--- a/ipc/namespace.c
+++ b/ipc/namespace.c
@@ -210,6 +210,7 @@ void put_ipc_ns(struct ipc_namespace *ns)
 			schedule_work(&free_ipc_work);
 	}
 }
+EXPORT_SYMBOL_GPL(put_ipc_ns);
 
 static struct ns_common *ipcns_get(struct task_struct *task)
 {
diff --git a/kernel/sched/syscalls.c b/kernel/sched/syscalls.c
index b215b0ead9a60ba1f8d1b5e7ee45e488e25ec901..cc6028938d0fb1e854b361eb4eb53c2b0d3b77b9 100644
--- a/kernel/sched/syscalls.c
+++ b/kernel/sched/syscalls.c
@@ -119,6 +119,7 @@ int can_nice(const struct task_struct *p, const int nice)
 {
 	return is_nice_reduction(p, nice) || capable(CAP_SYS_NICE);
 }
+EXPORT_SYMBOL_GPL(can_nice);
 
 #ifdef __ARCH_WANT_SYS_NICE
 
diff --git a/kernel/sched/wait.c b/kernel/sched/wait.c
index d033f600f48c6fc3a0a088ea5d9f6ed95ec4c86e..54b1a1b7ae4e8d0c785bec8959a0aaf0be904351 100644
--- a/kernel/sched/wait.c
+++ b/kernel/sched/wait.c
@@ -232,6 +232,7 @@ void __wake_up_pollfree(struct wait_queue_head *wq_head)
 	/* POLLFREE must have cleared the queue. */
 	WARN_ON_ONCE(waitqueue_active(wq_head));
 }
+EXPORT_SYMBOL_GPL(__wake_up_pollfree);
 
 /*
  * Note: we use "set_current_state()" _after_ the wait-queue add,
diff --git a/kernel/task_work.c b/kernel/task_work.c
index 0f7519f8e7c93f9a4536c26a341255799c320432..45fd146b85df4eb8bb14db22ccb84ad6d158317e 100644
--- a/kernel/task_work.c
+++ b/kernel/task_work.c
@@ -102,6 +102,7 @@ int task_work_add(struct task_struct *task, struct callback_head *work,
 
 	return 0;
 }
+EXPORT_SYMBOL_GPL(task_work_add);
 
 /**
  * task_work_cancel_match - cancel a pending work added by task_work_add()
diff --git a/mm/list_lru.c b/mm/list_lru.c
index 36662d02ff96310bccf7eb1c22cd166d8bc5bcf4..49305202038415a295c20dbe024d3b09368e2998 100644
--- a/mm/list_lru.c
+++ b/mm/list_lru.c
@@ -231,6 +231,7 @@ bool list_lru_add(struct list_lru *lru, struct list_head *item, int nid,
 	list_lru_unlock(l);
 	return ret;
 }
+EXPORT_SYMBOL_GPL(list_lru_del);
 
 bool list_lru_add_irq(struct list_lru *lru, struct list_head *item,
 		      int nid, struct mem_cgroup *memcg)
diff --git a/mm/memory.c b/mm/memory.c
index 8b0c2c735d3de76617dcf6d47239e92af15acc61..af2cbc54607c454df4ea00a53ee32c20d7e7d489 100644
--- a/mm/memory.c
+++ b/mm/memory.c
@@ -2378,6 +2378,7 @@ void zap_vma_range(struct vm_area_struct *vma, unsigned long address,
 	zap_vma_range_batched(&tlb, vma, address, size, NULL);
 	tlb_finish_mmu(&tlb);
 }
+EXPORT_SYMBOL_GPL(zap_vma_range);
 
 /**
  * zap_special_vma_range - zap all page table entries in a special vma range
diff --git a/mm/mmap_lock.c b/mm/mmap_lock.c
index 898c2ef1e9580371501b1bbc1ed5267286af2e15..c8ad31e9d7462abb6fd534e2ee222ee0e986bb30 100644
--- a/mm/mmap_lock.c
+++ b/mm/mmap_lock.c
@@ -431,6 +431,7 @@ struct vm_area_struct *lock_next_vma(struct mm_struct *mm,
 
 	return vma;
 }
+EXPORT_SYMBOL_GPL(lock_vma_under_rcu);
 #endif /* CONFIG_PER_VMA_LOCK */
 
 #ifdef CONFIG_LOCK_MM_AND_FIND_VMA
diff --git a/security/security.c b/security/security.c
index 2ee276ab15c5bcb55662a94fc9e654d9fa55619c..704b5d80377197c3c29cfc263e632cd50e93be8d 100644
--- a/security/security.c
+++ b/security/security.c
@@ -514,6 +514,7 @@ int security_binder_set_context_mgr(const struct cred *mgr)
 {
 	return call_int_hook(binder_set_context_mgr, mgr);
 }
+EXPORT_SYMBOL_GPL(security_binder_set_context_mgr);
 
 /**
  * security_binder_transaction() - Check if a binder transaction is allowed
@@ -529,6 +530,7 @@ int security_binder_transaction(const struct cred *from,
 {
 	return call_int_hook(binder_transaction, from, to);
 }
+EXPORT_SYMBOL_GPL(security_binder_transaction);
 
 /**
  * security_binder_transfer_binder() - Check if a binder transfer is allowed
@@ -544,6 +546,7 @@ int security_binder_transfer_binder(const struct cred *from,
 {
 	return call_int_hook(binder_transfer_binder, from, to);
 }
+EXPORT_SYMBOL_GPL(security_binder_transfer_binder);
 
 /**
  * security_binder_transfer_file() - Check if a binder file xfer is allowed
@@ -560,6 +563,7 @@ int security_binder_transfer_file(const struct cred *from,
 {
 	return call_int_hook(binder_transfer_file, from, to, file);
 }
+EXPORT_SYMBOL_GPL(security_binder_transfer_file);
 
 /**
  * security_ptrace_access_check() - Check if tracing is allowed
